### PR TITLE
New EIP to add precompile for BN256 hash-to-curve algorithms.

### PR DESCRIPTION
**Pull request by chgormanMH**
Summary:

This PR adds an EIP for precompile contracts for BN256 hash-to-curve algorithms.

### CODE REVIEW

File: EIPS/eip-BN256_HashToCurve.md
Comment (MicahZoltu) [inline EIPS/eip-BN256_HashToCurve.md:60]: By referring to 2537 inside this EIP you are creating a dependency on 2537 being final before this can go final.  I recommend removing that dependency, since you are explicitly wanting this EIP to be an alternative to 2537 IIUC.  When authoring EIPs, I recommend authoring them with a future reader in mind, rather than a historic reader.  Imagine in 5 or 10 years when someone is reading this EIP after it has been implemented and 2537 was not whether they are going to care at all about 2537.

File: EIPS/eip-BN256_HashToCurve.md
Comment (MicahZoltu) [inline EIPS/eip-BN256_HashToCurve.md:2]:
```suggestion
eip: 3068
```
Don't forget to rename this file to `EIPS/eip-3068.md` as well.

File: EIPS/eip-BN256_HashToCurve.md
Comment (MicahZoltu) [inline EIPS/eip-BN256_HashToCurve.md:81]: Is there a PDF version of this available that could be added to `../assets/eip-3068/weilsigs.pdf` and linked as such?  We try to avoid external links where possible as they tend to eventually disappear and we do not have the manpower to maintain dead links in EIPs.

File: EIPS/eip-BN256_HashToCurve.md
Comment (MicahZoltu) [inline EIPS/eip-BN256_HashToCurve.md:86]: Same comment as above, can we include this directly as an asset?

Comment (MicahZoltu) [reply EIPS/eip-BN256_HashToCurve.md:86]: Same comment for all of the other papers linked.

File: EIPS/eip-BN256_HashToCurve.md
Comment (MicahZoltu) [inline EIPS/eip-BN256_HashToCurve.md:123]: The specification should be *purely* technical, completely ignoring all history, sources, etc.  A good specification section looks more like code than an essay and just describes the spec.  If you want to talk about where things come from, or why decisions were made then Motivation or Rationale section are more appropriate.  Same goes for most of this paragraph.

The Rationale section is where you should put your reasons for making certain decisions within the specification.  For example, why you chose `80000 + 3 * len(bytes)` is very appropriate for the Rationale section, while the specification section should just define a constant or a function with little to know reasoning behind it.

The thing to keep in mind is that someone who just wants to implement the spec and doesn't care about the "why" at all (e.g., their boss told them to go implement EIP-3068) should be able to skip to the Specification section and immediately get down to business.

File: EIPS/eip-BN256_HashToCurve.md
Comment (MicahZoltu) [inline EIPS/eip-BN256_HashToCurve.md:143]: This is great content for the Rationale section (why you chose keccak256), but not great for the specification.

File: EIPS/eip-BN256_HashToCurve.md
Comment (MicahZoltu) [inline EIPS/eip-BN256_HashToCurve.md:141]: This is more of a personal recommendation rather than that of an editor, but rather than saying "`i` and `j` are bytes for domain separation", why not just use variable names in the pseudocode that aren't single letters long?  Perhaps just name the variables `domainSeparator1` and `domainSeparator2` or something?  Same with `p`, it seems silly to define it here as "the prime of the underlying field" and then refer to it throughout the code as `p`, why not just name it `fieldPrime` or something in the code so users don't have to use a lookup table?

Same feedback for basically all of the variables in this pseudocode (and really the same feedback for all mathematicians who write code 😬).

File: EIPS/eip-BN256_HashToCurve.md
Comment (MicahZoltu) [inline EIPS/eip-BN256_HashToCurve.md:301]: Since this is a cross client consensus change, it feels like there would be significant value to including test cases (inputs paired with expected output) so that clients can validate their implementation likely aligns with other clients prior to running integration tests.

This is not necessary for a draft EIP, but strongly advised for consensus EIPs at some point in the process.

File: EIPS/eip-BN256_HashToCurve.md
Comment (MicahZoltu) [reply EIPS/eip-BN256_HashToCurve.md:123]: Same basic content throughout the Specification section.

File: EIPS/eip-BN256_HashToCurve.md
Comment (MicahZoltu) [inline EIPS/eip-BN256_HashToCurve.md:4]:
```suggestion
author: Dr. Christopher Gorman MadHive (@chgormanMH)
```
This is likely going to break something, as the `,` character is special in the author line.  I'm not sure what `MadHive` is, but is this suggestion reasonable?

File: EIPS/eip-BN256_HashToCurve.md
Comment (MicahZoltu) [inline EIPS/eip-BN256_HashToCurve.md:310]:
```suggestion
TBD
```
In general, it is better just leave information out than include information that is *intended* to be deleted later.  By putting `TBD` in here it is less likely that it will get missed in final review if this section is never updated.

Also, keep in mind that the Implementation section isn't required for an EIP, so you can just leave the whole section out until you have one ready to be included.

File: EIPS/eip-BN256_HashToCurve.md
Comment (MicahZoltu) [inline EIPS/eip-BN256_HashToCurve.md:24]: Simple Summary should be much shorter.  Imagine an email subject line, or a forum sub-title, or a news headline or sub-headline.  Usually about a sentence is good, and often that sentence will be particularly terse and maybe not even grammatically correct!

It should not go into the "why" for the EIP, just describe in very pithy terms what this EIP is.  In this case, something like this may be good? (you know the matter better than I do to evaluate what a good headline would be):
```suggestion
Defines a hash-to-curve precompile for use in BLS12-381 ECC.
```

**Review by MicahZoltu — CHANGES_REQUESTED** [review]
The frontmatter stuff is most critical.  I do recommend addressing some of the other things, but it doesn't have to be done before merging as a draft (while the frontmatter changes are required).

File: EIPS/eip-BN256_HashToCurve.md
Comment (chgormanMH) [reply EIPS/eip-BN256_HashToCurve.md]: I have removed all references to EIP-2537 from the EIP.

File: EIPS/eip-BN256_HashToCurve.md
Comment (chgormanMH) [reply EIPS/eip-BN256_HashToCurve.md]: I renamed the file and updated it to EIP-3068.

File: EIPS/eip-BN256_HashToCurve.md
Comment (chgormanMH) [reply EIPS/eip-BN256_HashToCurve.md]: I have made local copies of this paper.

File: EIPS/eip-BN256_HashToCurve.md
Comment (chgormanMH) [reply EIPS/eip-BN256_HashToCurve.md]: Commentary in Specification has been reduced.

File: EIPS/eip-BN256_HashToCurve.md
Comment (chgormanMH) [reply EIPS/eip-BN256_HashToCurve.md]: Comment has been removed from Specification.

File: EIPS/eip-BN256_HashToCurve.md
Comment (chgormanMH) [reply EIPS/eip-BN256_HashToCurve.md]: Algorithms have been updated with variable names, hopefully resulting in pseudocode that is more understandable.

File: EIPS/eip-BN256_HashToCurve.md
Comment (chgormanMH) [reply EIPS/eip-BN256_HashToCurve.md]: I have removed `, MadHive`. I work at MadHive. Is there a standard practice of including one's employer? I developed this work during my employment.

File: EIPS/eip-BN256_HashToCurve.md
Comment (chgormanMH) [reply EIPS/eip-BN256_HashToCurve.md]: Implementation section has been reduced to TBD.

File: EIPS/eip-BN256_HashToCurve.md
Comment (chgormanMH) [reply EIPS/eip-BN256_HashToCurve.md]: Simple Summary has been simplified.

File: EIPS/eip-BN256_HashToCurve.md
Comment (chgormanMH) [reply EIPS/eip-BN256_HashToCurve.md]: This link has been updated and a local copy of the paper has been included.

File: EIPS/eip-BN256_HashToCurve.md
Comment (chgormanMH) [reply EIPS/eip-BN256_HashToCurve.md]: I am surprised this would be "considered a cross client consensus change". I believe the hash-to-curve algorithms would not touch consensus at all, as I am only looking to add precompile contracts to the EVM. Am I mistaken or misunderstanding something?

If specific test cases would be included (such as known results for BaseToG1 or BaseToG2), those could be included without much hassle. Is this what you are suggesting?

Comment (MicahZoltu) [reply EIPS/eip-BN256_HashToCurve.md]: New precompiles are a consensus change in that if clients don't agree on the output of a precompile given ea particular input then they will desync and we'll no longer have consensus across clients.

File: EIPS/eip-BN256_HashToCurve.md
Comment (MicahZoltu) [reply EIPS/eip-BN256_HashToCurve.md]: EIPs are all CC0, so there is no attribution.  The author line is mainly just for access control on who can freely make changes to the EIP leading up to final.  The author line is left in for final EIPs, but it is pretty meaningless given the license.

File: EIPS/eip-BN256_HashToCurve.md
Comment (chgormanMH) [reply EIPS/eip-BN256_HashToCurve.md]: MadNet paper has been added. This should take care of all references.

File: EIPS/eip-BN256_HashToCurve.md
Comment (chgormanMH) [reply EIPS/eip-BN256_HashToCurve.md]: I will add some test cases in the near future. Yes, having consistent outputs across clients is critical and I now understand why tests are needed.

File: EIPS/eip-BN256_HashToCurve.md
Comment (chgormanMH) [reply EIPS/eip-BN256_HashToCurve.md]: Not a problem, then.

File: EIPS/eip-3068.md
Comment (MicahZoltu) [inline EIPS/eip-3068.md:282]:
```suggestion
[whitepaper](../assets/eip-3068/madnet.pdf),
```

File: EIPS/eip-3068.md
Comment (MicahZoltu) [inline EIPS/eip-3068.md:10]:
```suggestion
requires: 198, 1108
```

File: EIPS/eip-3068.md
Comment (MicahZoltu) [inline EIPS/eip-3068.md:67]: This doesn't appear to be a human readable document.  Also, it completely chokes the spellchecker that is blocking merging this EIP as draft.  Can a human readable version be included instead?

**Review by MicahZoltu — CHANGES_REQUESTED** [review]
CI is blocked by these issues (at least, may be more buried once those get fixed).

File: EIPS/eip-3068.md
Comment (chgormanMH) [reply EIPS/eip-3068.md]: I converted the PostScript file to PDF. Hopefully that fixes things.

File: EIPS/eip-3068.md
Comment (chgormanMH) [reply EIPS/eip-3068.md]: Fixed.

File: EIPS/eip-3068.md
Comment (chgormanMH) [reply EIPS/eip-3068.md]: Good catch. Fixed.

**Review by MicahZoltu — APPROVED** [review]

**Pull request merged**
